src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -1,32 +1,165 @@
-                     PageSpan)
+                     PageSpan (..))
 import qualified Database.LSMTree.Internal.Run.Index.Compact as Index
 import           Database.LSMTree.Internal.Serialise
+import           System.FS.API (Handle)
+import           System.FS.BlockIO.API
 
 -- | TODO: placeholder type for a run, replace by actual type once implemented
-type Run fd = (fd, Bloom SerialisedKey, CompactIndex)
+type Run h = (h, Bloom SerialisedKey, CompactIndex)
+
+rBloom :: Run h -> Bloom SerialisedKey
+rBloom (_, b, _) = b
+
+rCindex :: Run h -> CompactIndex
+rCindex (_, _, c) = c
+
+rHandle :: Run h -> h
+rHandle (h, _, _) = h
+
+-- | Prepare disk lookups by doing bloom filter queries, index searches and
+-- creating 'IOOp's. The result is a vector of 'IOOp's and a vector of indexes,
+-- both of which are the same length. The indexes record the run and key
+-- associated with each 'IOOp'.
+prepLookups ::
+     PrimMonad m
+  => V.Vector (Run (Handle h))
+  -> V.Vector SerialisedKey
+  -> m (VU.Vector (RunIx, KeyIx), V.Vector (IOOp m h))
+prepLookups rs ks = do
+  let rkixs = bloomQueries rs ks
+  ioops <- indexSearches rs ks rkixs
+  pure (rkixs, ioops)
 
--- | Prepare disk lookups by doing bloom filter queries and index searches.
+type KeyIx = Int
+type RunIx = Int
+type ResIx = Int -- Result index
+
+-- | Perform a batch of bloom queries. The result is a tuple of indexes into the
+-- vector of runs and vector of keys respectively.
 --
--- Note: results are grouped by key instead of file descriptor, because this
--- means that results for a single key are close together.
-prepLookups :: [Run fd] -> [SerialisedKey] -> [(SerialisedKey, (fd, PageSpan))]
-prepLookups runs ks =
-    [ (k, (fd, pspan))
-    | k <- ks
-    , r@(fd,_,_) <- runs
-    , pspan <- toList (prepLookup r k)
-    ]
-
-prepLookup :: Run fd -> SerialisedKey -> Maybe PageSpan
-prepLookup (_fd, b, fpix) k
-  | Bloom.elem k b = Index.toPageSpan $ Index.search k fpix
-  | otherwise      = Nothing
+-- TODO: we consider it likely that we could implement a optimised, batched
+-- version of bloom filter queries, which would largely replace this function.
+bloomQueries :: V.Vector (Run a) -> V.Vector SerialisedKey -> VU.Vector (RunIx, KeyIx)
+bloomQueries rs ks = VU.create $ do
+      -- The result vector can be of variable length, so we use a generous
+      -- estimate here, and we grow the vector if needed. TODO: tune the
+      -- starting estimate based on the expected true- and false-positives.
+      res <- VUM.unsafeNew (2 * ksN)
+      resixVar <- newPrimVar 0
+      res' <- go res resixVar
+      resix <- readPrimVar resixVar
+      -- The result vector is not guaranteed to be full, so we slice it.
+      pure $ VUM.slice 0 resix res'
+  where
+    !rsN = V.length rs
+    !ksN = V.length ks
+
+    go :: forall s.
+         VUM.MVector s (RunIx, KeyIx)
+      -> PrimVar s ResIx
+      -> ST s (VUM.MVector s (RunIx, KeyIx))
+    go res0 resixVar = loop1 res0 0
```

Comment:
The nesting was intentional to not have to thread through values that don't change within a loop. I can give it a try, though I'm not sure how GHC would optimise it

IMO the nesting is less error-prone than threading "constant" parameters through

Reply:
Yes, it's ok (and often faster) to have fewer nestings and just pass in args like `res` even if those args are not modified at all in the recursive calls.

That said, I don't think we need to worry too much about optimising `bloomQueries` right now because the plan is to write a special bulk version of it that does all the prefetching. So we can do micro-optimisations then.

Reply:
Okay, I've made into a mono-loop